src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,189 @@
+    const state = initialize(digest_length);
+    update(state, Bytes.from(data).bytes);
+    const out = final(state);
+    return Bytes.from(out);
+  },
+  get IV() {
+    return BLAKE2BConstants.IV.map((x) => UInt64.from(x));
+  },
+};
+
+function initialize(outlen: number): {
+  h: UInt64[];
+  t: UInt64[];
+  f: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+} {
+  const h = BLAKE2B.IV.slice(); // shallow copy IV to h
+  h[0] = UInt64.from(0x01010000).xor(UInt64.from(outlen)).xor(h[0]);
+  return {
+    h,
+    t: [UInt64.zero, UInt64.zero],
+    f: [UInt64.zero, UInt64.zero],
+    buf: [],
+    buflen: 0,
+    outlen,
+  };
+}
+
+function G(
+  v: UInt64[],
+  a: number,
+  b: number,
+  c: number,
+  d: number,
+  x: UInt64,
+  y: UInt64
+) {
+  v[a] = v[a].addMod64(v[b].addMod64(x));
+  v[d] = v[d].xor(v[a]).rotate(32, 'right');
+
+  v[c] = v[c].addMod64(v[d]);
+  v[b] = v[b].xor(v[c]).rotate(24, 'right');
+
+  v[a] = v[a].addMod64(v[b].addMod64(y));
+  v[d] = v[d].xor(v[a]).rotate(16, 'right');
+
+  v[c] = v[c].addMod64(v[d]);
+  v[b] = v[b].xor(v[c]).rotate(63, 'right');
+}
+
+function compress(state: {
+  h: UInt64[];
+  t: UInt64[];
+  f: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+}): void {
+  const { h, t, f, buf } = state;
+  const v = h.concat(BLAKE2B.IV); // initalize local work vector. First half from state and second half from IV.
+
+  v[12] = v[12].xor(t[0]); // low word of the offset
+  v[13] = v[13].xor(t[1]); // high word of the offset
+  v[14] = v[14].xor(f[0]);
+  v[15] = v[15].xor(f[1]);
+
+  const m: UInt64[] = [];
+  for (let i = 0; i < 16; i++) {
+    m.push(
+      UInt64.from(
+        buf[i * 8]
+          .toUInt64()
+          .or(buf[i * 8 + 1].toUInt64().leftShift(8))
+          .or(buf[i * 8 + 2].toUInt64().leftShift(16))
+          .or(buf[i * 8 + 3].toUInt64().leftShift(24))
+          .or(buf[i * 8 + 4].toUInt64().leftShift(32))
+          .or(buf[i * 8 + 5].toUInt64().leftShift(40))
+          .or(buf[i * 8 + 6].toUInt64().leftShift(48))
+          .or(buf[i * 8 + 7].toUInt64().leftShift(56))
+      )
+    );
+  }
+
+  for (let i = 0; i < 12; i++) {
+    const s = BLAKE2BConstants.SIGMA[i % 10];
+    G(v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+    G(v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+    G(v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+    G(v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+    G(v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+    G(v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+    G(v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+    G(v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+  }
+
+  for (let i = 0; i < 8; i++) {
+    h[i] = v[i].xor(v[i + 8]).xor(h[i]);
+  }
+}
+
+function update(
+  state: {
+    h: UInt64[];
+    t: UInt64[];
+    f: UInt64[];
+    buf: UInt8[];
+    buflen: number;
+    outlen: number;
+  },
+  input: UInt8[]
+): void {
+  for (let i = 0; i < input.length; i++) {
+    if (state.buflen === 128) {
+      state.t[0] = state.t[0].add(128); // ?
```

Comment:
what's the question mark for?

Reply:
I left that while debugging and forgot to remove it.